src/modules/flow/http-client/http-client.c

```diff
@@ -1384,4 +1395,286 @@ get_key_process(struct sol_flow_node *node, void *data,
     return 0;
 }
 
+static void
+create_url_close(struct sol_flow_node *node, void *data)
+{
+    struct create_url_data *mdata = data;
+
+    free(mdata->protocol);
+    free(mdata->server);
+    free(mdata->path);
+    free(mdata->fragment);
+    sol_http_param_free(&mdata->params);
+    sol_arena_del(mdata->arena);
+}
+
+static int
+add_query(struct sol_arena *arena, struct sol_http_param *params,
+    const struct sol_str_slice key, const struct sol_str_slice value)
+{
+    char *cvalue, *ckey;
+
+    ckey = sol_arena_strdup_slice(arena, key);
+    SOL_NULL_CHECK(ckey, -ENOMEM);
+
+    cvalue = sol_arena_strdup_slice(arena, value);
+    SOL_NULL_CHECK(cvalue, -ENOMEM);
+
+    if (!sol_http_param_add(params,
+        SOL_HTTP_REQUEST_PARAM_QUERY(ckey, cvalue))) {
+        SOL_ERR("Could not add the HTTP param %s:%s", ckey, cvalue);
+        return -ENOMEM;
+    }
+
+    return 0;
+}
+
+static int
+parse_query(struct sol_arena *arena, const char *query,
+    struct sol_http_param *params)
+{
+    struct sol_vector tokens;
+    struct sol_str_slice *slice, key, value;
+    uint16_t i;
+    char *sep;
+    int r = 0;
+
+    tokens = sol_util_str_split(sol_str_slice_from_str(query), "&", 0);
+
+    SOL_VECTOR_FOREACH_IDX (&tokens, slice, i) {
+        sep = memchr(slice->data, '=', slice->len);
+
+        if (!sep) {
+            SOL_ERR("Could not find the '=' separator at:%.*s",
+                SOL_STR_SLICE_PRINT(*slice));
+            r = -EINVAL;
+            break;
+        }
+
+        key.data = slice->data;
+        key.len = sep - slice->data;
+
+        value.data = sep + 1;
+        value.len = slice->len - key.len - 1;
+
+        r = add_query(arena, params, key, value);
+        SOL_INT_CHECK_GOTO(r, < 0, exit);
+    }
+
+exit:
+    sol_vector_clear(&tokens);
+    return r;
+}
+
+static int
+replace_uri(struct create_url_data *mdata, const char *uri)
+{
+    struct sol_str_slice protocol, server, path, fragment;
+    int port, r;
+
+    r = sol_http_split_uri(uri, &protocol, &server,
+        &path, &fragment, &mdata->params, &port);
+    SOL_INT_CHECK(r, < 0, r);
+
+    r = sol_util_replace_str_from_slice_if_changed(&mdata->protocol, protocol);
+    SOL_INT_CHECK(r, < 0, r);
+
+    r = sol_util_replace_str_from_slice_if_changed(&mdata->server, server);
+    SOL_INT_CHECK(r, < 0, r);
+
+    if (fragment.len) {
+        r = sol_util_replace_str_from_slice_if_changed(&mdata->fragment,
```

Comment:
you should always replace fragment... otherwise getting an empty fragment after a non-empty one will keep the old (and incorrect). Same for path.